### PR TITLE
Add SignerProvider field to jwt.Config

### PR DESCRIPTION
**Pull request by djrollins**
Add SignerProvider field to jwt.Config which, if set, is used to create a jws.Signer from the Config.PrivateKeyID field which is then used sign the JWT instead of using the provided Config.PrivateKey. This allows the library to easily be integrated with external key management tools.

### CODE REVIEW

File: jwt/jwt.go [hunk 1]
```diff
@@ -40,12 +41,18 @@ type Config struct {
...
 	// Use the following command to convert a PKCS 12 file into a PEM.
 	//
 	//    $ openssl pkcs12 -in key.p12 -out key.pem -nodes
 	//
 	PrivateKey []byte
 
+	// SignerProvider is a function that is used to create a jws.Signer from the
+	// PrivateKeyID which is then used to sign JWT payloads. This takes
+	// precedence over default signer using the PrivateKey.
+	SignerProvider func(PrivateKeyID string) (jws.Signer, error)
```
Comment (a-h) [inline jwt/jwt.go:54]:
```suggestion
	SignerProvider func(privateKeyID string) (jws.Signer, error)
```

File: jwt/jwt.go [hunk 2]
```diff
@@ -124,7 +127,17 @@ func (js jwtSource) Token() (*oauth2.Token, error) {
 	}
 	h := *defaultHeader
 	h.KeyID = js.conf.PrivateKeyID
-	payload, err := jws.Encode(&h, claimSet, pk)
+	payload := ""
+	var err error = nil
+	if js.conf.SignerProvider == nil {
+		var pk *rsa.PrivateKey
+		pk, err = internal.ParseKey(js.conf.PrivateKey)
```
Comment (a-h) [inline jwt/jwt.go:134]: Looks like you forgot to return if there's an error here.

```suggestion
		pk, err = internal.ParseKey(js.conf.PrivateKey)
		if err != nil {
			return nil, err
		}
```

File: jwt/jwt.go [hunk 3]
```diff
@@ -124,7 +127,17 @@ func (js jwtSource) Token() (*oauth2.Token, error) {
...
+		payload, err = jws.Encode(&h, claimSet, pk)
+	} else {
+		var signer jws.Signer
+		signer, err = js.conf.SignerProvider(js.conf.PrivateKeyID)
```
Comment (a-h) [inline jwt/jwt.go:138]: Same here.

```suggestion
		signer, err = js.conf.SignerProvider(js.conf.PrivateKeyID)
		if err != nil {
			return nil, err
		}
```

File: jwt/jwt_test.go [hunk 4]
```diff
@@ -232,6 +237,25 @@ func TestJWTFetch_AssertionPayload(t *testing.T) {
 				"private1": "claim1",
 			},
 		},
+		{
+			Email: "aaa2@xxx.com",
+			SignerProvider: func(_ string) (jws.Signer, error) {
+				key, err := internal.ParseKey(dummyPrivateKey)
+
```
Comment (a-h) [inline jwt/jwt_test.go:244]: Excess line.

```suggestion
```

File: jwt/jwt_test.go [hunk 5]
```diff
@@ -232,6 +237,25 @@ func TestJWTFetch_AssertionPayload(t *testing.T) {
...
+				if err != nil {
+					return nil, err
+				}
+
+				return func(data []byte) (sig []byte, err error) {
+					h := sha256.New()
+					h.Write(data)
+					return rsa.SignPKCS1v15(rand.Reader, key, crypto.SHA256, h.Sum(nil))
+				}, err
+
```
Comment (a-h) [inline jwt/jwt_test.go:254]:
```suggestion
```

File: jwt/jwt.go (hunk 3 above)
Comment (djrollins) [reply jwt/jwt.go]: Nice spot!

File: jwt/jwt.go [hunk 6]
```diff
@@ -124,7 +127,23 @@ func (js jwtSource) Token() (*oauth2.Token, error) {
...
+		}
+		payload, err = jws.Encode(&h, claimSet, pk)
+	} else {
+		var signer jws.Signer
+		signer, err = js.conf.SignerProvider(h.KeyID)
+		if err != nil {
+			return nil, err
+		}
+		payload, err = jws.EncodeWithSigner(&h, claimSet, signer)
+	}
```
Comment (djrollins) [inline jwt/jwt.go:146]: I toyed with the idea of this to get around the awkward scoping of `err`, but I don't know it's particularly idiomatic Go... Thoughts 
```suggestion
	var encode func(*jws.Header, *jws.ClaimSet) (string, error)
	if js.conf.SignerProvider == nil {
		encode = func(h *jws.Header, c *jws.ClaimSet) (string, error) {
			pk, err := internal.ParseKey(js.conf.PrivateKey)
			if err != nil {
				return "", err
			}
			return jws.Encode(h, c, pk)
		}
	} else {
		encode = func(h *jws.Header, c *jws.ClaimSet) (string, error) {
			signer, err := js.conf.SignerProvider(h.KeyID)
			if err != nil {
				return "", err
			}
			return jws.EncodeWithSigner(h, c, signer)
		}
	}
	payload, err := encode(&h, claimSet)
```

Comment (djrollins) [reply jwt/jwt.go:146]: @a-h

**Pull request closed without merging**